### PR TITLE
Move registration logic to new controller

### LINKED ISSUE #27526: Simpler and friendly URL of page with Registration form
### Prerequisites

- [X] I understand and accept the project's [code of conduct](https://github.com/PrestaShop/PrestaShop/blob/develop/CODE_OF_CONDUCT.md)
- [X] I have already [searched in existing features request](https://github.com/PrestaShop/PrestaShop/issues?q=is%3Aissue+label%3AImprovement%2CFeature%2CDeveloper+Feature+) and found no previous suggestion of this feature.

### Is your feature request related to a problem?

Is there any really good reason why the page with Registration form is calling by following Friendly URL?

`/authenticating?create_account=1`

Why is calling page with Login form with additional ?create_account=1 variable?

What about having nice and short URL instead?

`/sign-up`  or `/register` or `/new-account` or whatever merchant set in back office.

A few advantages:
* the merchant can choose to allow to indexing page with login form but not with registration form, and vice versa
* URL can have tracking or affiliate marketing parameters
* fully translatable friendly URL without english only parameter
* in some other languages we don't want use "authenticating" phrase anywhere in the eshop because in context it could meaning something quite different or is too formal in the particular country and culture. We want to use "register" or "new-account" in the URL. And now it is quite funny using friendy URL where URL to registration form include phrase "login".

### Describe the solution you'd like

Could you please implement this improvement if there is no reason to have ?create_account=1 in the URL?

### Alternatives you've considered

_No response_

### Additional context

_No response_

### Do you plan to work on this feature?

- [ ] I'm willing to contribute a [formal specification](https://github.com/PrestaShop/prestashop-specs).
- [ ] I'm willing to provide any wireframes or design assets required for this feature.
- [ ] I'm willing to submit a Pull Request that implements this feature.
- [ ] I'm willing to help verify that the implemented feature works as intended and produces no unintended side effects.

### PR DESCRIPTION
**Pull request by Hlavtox**
| Questions         | Answers
| ----------------- | -------------------------------------------------------
| Branch?           | develop
| Description?      | This PR moves the registration logic from mixed auth controller. This controller was practically a huge IF/ELSE block. User can now properly translate the URL, it will be cleaner and you can set if you want to have it in sitemap for example.
| Type?             | improvement
| Category?         | FO
| BC breaks?        | See below.
| Deprecations?     | np
| Fixed ticket?     | Fixes #27526
| Related PRs     |   https://github.com/PrestaShop/PrestaShop/pull/30093
| How to test?      | See below. Needs https://github.com/PrestaShop/classic-theme/pull/33 to fix width issue.
| Possible impacts? | Only login and registration in front office. No other changes.

**BC breaks**
- ✅ There is no BC break for themes, it uses still the same templates and URLs a URLs for login/register pages are passed from the core.
- ✅ I added a redirect to the new controller, if somebody linked to the old controller or does a GET request to it.
- 🟡  The new controller will not be accounted for in sitemap module for example, but I think it's not particullary significant.
- 🔴 BC break will be, if somebody did a POST request with the registration logic to the old controller - for some reason.

**How to test**
You can test the PR on already installed Prestashop, but you need to:
- Run `composer dump-autoload`.
- Go to SEO&URL settings and add the new registration controller to your URL list.
- Delete `var/cache/prod/class_index.php` and `var/cache/dev/class_index.php.`

**BC break can be further reduced by**
```
    public function getPageLink($controller, $ssl = null, $idLang = null, $request = null, $requestUrlEncode = false, $idShop = null, $relativeProtocol = false)
    {
        if ($controller == 'authentication' && isset($request['create_account'])) {
            unset($request['create_account']);
            return $this->getPageLink('registration', $ssl, $idLang, $request, $requestUrlEncode, $idShop, $relativeProtocol);
        }
```

---
This change is [Reviewable](https://reviewable.io/reviews/prestashop/prestashop/27755)

### CODE REVIEW

**Comment by kpodemski** [comment]
It's a massive BC Break. I don't think it's worth it.

Edit:
- basically, every module that has an OPC feature
- every link that points to the registration form
- every module that has a "popup registration", etc.

And it's only a slight improvement. Maybe v8 is suitable for this kind of change, but we're increasing the scope of migration again.

**Comment by Hlavtox** [comment]
@kpodemski Redirect added 😛

**Review by matks — COMMENTED** [review]
Just like @kpodemski I am bit afraid of such a big change suddenly.

If we had deprecated the previous controller before, adding warnings "Hello! this way of doing the registration will not work in next version" it would be OK but here this change is immediate and without early warning.

As mentioned by @kpodemski we try to reassure people about the jump from 1.7 to 8 😅 and so far we kept changes on BO. This BC break in FO might be overkill.

Your code is great 😉 the problem is how to introduce it, we need the right timing.

Wdyt @PrestaShop/prestashop-maintainers

**Comment by Hlavtox** [comment]
@matks I absolutely understand your concerns, but if I also implement the backward compatibility fix to `Link` function, this change would only affect people who for some reason hardcode `/login` in to their code.

Soooo, maybe it won't be that intensive?

**Comment by kpodemski** [comment]
Hey @Hlavtox 

I'd put it on hold and plan it for v9. It really doesn't change the quality of the software. It has an additional value, as everything that comes from you @Hlavtox, but I think it's just not worth it for v8. If you need to add workarounds (Like these checks in `Link` that we talked about) to prevent BC Breaks, and the change is not that crucial, it simply is not worth it.

**Review by matks — COMMENTED** [review]
I agree with @kpodemski 

We are really worried about making the upgrade from 1.7 to 8 **not scary**. We want people to be OK, we want people to upgrade from 1.7 to 8 and think "ok, it was not that bad, it was better than 1.6 to 1.7". And to do this we need to restrict ourselves to backward-compatible changes and/or small changes (not backward compatible), excluding the topics we need to push like product page v2.

File: controllers/front/AuthController.php
Comment (Progi1984) [inline controllers/front/AuthController.php:58]: Could be an unique `if`

File: controllers/front/RegistrationController.php
Comment (Progi1984) [inline controllers/front/RegistrationController.php:26]:
```suggestion
 */

declare(strict_types=1);

class RegistrationControllerCore extends FrontController
```

File: controllers/front/RegistrationController.php
Comment (Progi1984) [inline controllers/front/RegistrationController.php]: Please add PHPDoc

File: controllers/front/RegistrationController.php
Comment (Progi1984) [inline controllers/front/RegistrationController.php:57]:
```suggestion
                static function (bool $carry, bool $item): bool {
```

**Review by Progi1984 — COMMENTED** [review]
Some small feedbacks but it's a good refacto.

File: controllers/front/RegistrationController.php
Comment (Hlavtox) [reply controllers/front/RegistrationController.php]: @Progi1984 It's already documented in `classes/controller/FrontController`, is it needed to PHPDoc it in every controller that inherits from FrontController? It would be like 20 times the same PHPDoc for each property.

**Review by Progi1984 — CHANGES_REQUESTED** [review]
Hello @Hlavtox, thanks for your contribution, would you mind to rebase your pull request? Thanks

Reference : https://anavarre.net/how-to-rebase-a-github-pull-request/

**Comment by kpodemski** [comment]
Maybe I'll ping @iqit-commerce to comment on this change. He might be a good person to provide us with feedback from the theme developer's perspective.

**Review by kpodemski — CHANGES_REQUESTED** [review]
With all these changes that we have in the core because of the removed deprecations, I think we can go for it, but first it needs a rebase :D

File: controllers/front/AuthController.php
Comment (jolelievre) [inline controllers/front/AuthController.php:48]: I guess the only thing that is left to handle, as we discussed in demo, is to handle this redirection with POST correctly right?

Comment (jolelievre) [reply controllers/front/AuthController.php:48]: So here are a few solutions amongst the one we discussed:
- perform a redirection with appropriate HTTP code to avoid losing POST method, not sure it works based on https://stackoverflow.com/questions/2604530/a-good-way-to-redirect-with-a-post-request But you can test if 301 or 302 works, seems like 307 should be the best candidate but depending on browsers it can remove the POST parameters..
- if the HTTP code doesn't work perform an internal forward using the new controller (see suggestion below)

Comment (jolelievre) [reply controllers/front/AuthController.php:48]:
```suggestion
            $controller = new RegistrationController();
            
            return $tontroller->initContent();
```

I'm not sure AT ALL if this implementation is correct or even functional 😅. but it gives the general idea

**Review by jolelievre — COMMENTED** [review]
Thanks @Hlavtox

File: controllers/front/AuthController.php
Comment (Hlavtox) [reply controllers/front/AuthController.php:48]: @jolelievre This redirection works perfectly. We were talking about a situation when you do:
- `$this->context->link->getPageLink('authentication', true, null, ['create_account' => '1']);`
- Pass this URL to smarty or JS for your login popup.
- Make a POST request to it.

We could fix it by returning a correct URL in getPageLink, WDYT?

Then the only way it could fail is, if somebody was POSTing to hardcoded `/login?create_account` for whatever reason. 😄

Comment (jolelievre) [reply controllers/front/AuthController.php:48]: Actually, the case with hardcoded URL is exactly what I had in mind In this case we need to handle the POST redirection So we are sure that this modification from your PR is fully backward compatible

And this way no need to handle a tricky edge case in `getPageLink`

**Comment by iqit-commerce** [comment]
> Maybe I'll ping @iqit-commerce to comment on this change. He might be a good person to provide us with feedback from the theme developer's perspective.


Hello,

From theme perspective view I do not see any problem with PR. I think it may only impact on some social login modules or modules used to extend registration form.

**Comment by Hlavtox** [comment]
Rebased, fixed breadcrumbs, commented, improved 'how to test'.

**Comment by Hlavtox** [comment]
@kpodemski Do we go for this?

**Review by kpodemski — DISMISSED** [review]

**Review by Progi1984 — DISMISSED** [review]

**Comment by AureRita** [comment]
Hi @Hlavtox ,

Thank you for your PR, I tested it and, as we want to, the link for the registration change as we can see : 

with your pr :
image

without your pr : 
image

but it also changes the text input size of the registration, as we can see : 

with your pr 
image

without your pr : 
image

so that create some fails on automated tests, @nesrineabdmouleh will give the last approval 

Thanks

**Comment by Hlavtox** [comment]
@NeOMakinG Bro can you check the width thing? I sense some underground magic in CSS 😁

**Comment by nesrineabdmouleh** [comment]
Hello,

Thank you @Hlavtox for your PR and @AureRita for your tests and for the ping!
The problem of automated tests that the title of 'FO>Create account' page is changed from 'Login' to 'ShopName' 

I will add a commit to fix it.
Many thanks!

**Review by boubkerbribri — DISMISSED** [review]
LGTM for UI tests

**Review by boubkerbribri — APPROVED** [review]

**Comment by NeOMakinG** [comment]
@Hlavtox the form width will be fixed by https://github.com/PrestaShop/classic-theme/pull/33

**Comment by Hlavtox** [comment]
@NeOMakinG Thanks bro!!!

@AureRita All should be good now. 👍

**Review by kpodemski — APPROVED** [review]

**Review by AureRita — APPROVED** [review]
Hi @Hlavtox 

Indeed, all it's good ffor me, 

Thank you for your PR

**Comment by kpodemski** [comment]
Thanks, @Hlavtox & @AureRita ❤️ 

Daniel, maybe you'd like to present this next week on Public Demo?

**Comment by Caleydon** [comment]
Will be updated also Link list module in order to allow add new "Registration" link?
https://github.com/PrestaShop/ps_linklist

**Comment by NeOMakinG** [comment]
> https://github.com/PrestaShop/ps_linklist

It will require a PR from someone probably, if you want to do it, feel free!

Or create an issue too

**Comment by jf-viguier** [comment]
The registration and authentication cannot be in the same page anymore with this PR. In 1.6 in was in the same page, so it's a regression from 1.6 I guess.

**Comment by Caleydon** [comment]
It would be nice to add meta data for new Registration controller by the installation process

File: /install/data/xml/meta.xml

```
<meta id="registration">
  <page>registration</page>
  <configurable>1</configurable>
</meta>
```

File: /install/langs/en/data/meta.xml

```
<meta id="registration" id_shop="1">
  <title>Registration</title>
  <description>Create your new customer account in our eshop.</description>
  <keywords>customer, registration<keywords/>
  <url_rewrite>registration</url_rewrite>
</meta>
```

**Pull request merged**
